### DISCUSSION TITLE
how to make forecasts like OLS using DirectReductionForecaster?

### DISCUSSION THREAD

**Discussion by sakaiando** [alan-turing-institute/sktime#7032] [Q&A] [answered]
I thought DirectReductionForecaster can mimic the behavior of sklearn LinearRegression by setting window_length = 0, but having the same X_pred doesn't lead to the same forecasts, probably because different models are fit in different horizon, which I feel is unnecessarily throwing away observations.

How can I let DirectReductionForecaster do the same thing as LinearRegression?

```
import pandas as pd
import numpy as np
from sktime.datasets import load_longley
from sktime.forecasting.compose import DirectReductionForecaster
from sklearn.linear_model import LinearRegression

y, X = load_longley()
forecaster = DirectReductionForecaster(estimator=LinearRegression(fit_intercept=False),
                                       window_length = 0)
model = LinearRegression(fit_intercept=False)

# forecaster is OLS of y on X when fh=0, but not when fh=1,2,...
print(forecaster.fit(y,X,fh=[0,1,2,3]).get_fitted_params()['estimators'][0].coef_ - model.fit(X,y).coef_, 'all 0')
print(forecaster.fit(y,X,fh=[0,1,2,3]).get_fitted_params()['estimators'][1].coef_ - model.fit(X.iloc[1:,:],y.iloc[1:]).coef_, 'all 0')
print(forecaster.fit(y,X,fh=[0,1,2,3]).get_fitted_params()['estimators'][2].coef_ - model.fit(X.iloc[2:,:],y.iloc[2:]).coef_, 'all 0')

# How to make the same forecast for all horizons when X_pred are the same?
X_pred = pd.DataFrame(X.iloc[-3:,:].values*0,index=X.index[-3:]+3,columns=X.columns)
X_pred.iloc[:,-2] = 100
y_fh = forecaster.fit(y=y,X=X,fh=[0,1,2,3]).predict(X=X_pred)
print(X_pred, 'same across horizon')
print(y_fh,'not the same across horizon')
print(model.fit(X,y).predict(X_pred),'I want this behavior')
```

The fit() method doesn't allow an empty fh, so I can't use something like forecaster.fit(y=y,X=X).predict(X=X_pred,fh=[0,1,2,3]).

**Comment by ericjb** [comment]
If X, y, and X_pred are as in your code and then you execute

```
model = LinearRegression(fit_intercept=False)
model.fit(X,y)
model.predict(X_pred)
```

the output is

```
array([-57.93203636, -57.93203636, -57.93203636])
```


This output does not agree with what your code outputs next to the comment "I want this behavior".
Please confirm which is the output you "want".

**Reply by sakaiando** [reply]
Thanks @ericjb, I fixed the last line.

As you point out, I want

`array([-57.93203636, -57.93203636, -57.93203636])`

but yh is

```
Period
1962    71375.013163
1963      -56.000224
1964      -90.812194
1965      -95.138320
```

Basically, I want to use LinearRegression in ForecastingGridSearchCV, but DirectReductionForecaster changes the behavior, so if you know how to achieve the objective, please kindly let me know.

**Comment by ericjb** [comment, marked as answer]
I think you want the YfromX forecaster (it is found in the same file as DirectReductionForecaster)

```
from sktime.forecasting.compose import YfromX
f = YfromX(LinearRegression(fit_intercept=False))
f.fit(y=y, X=X, fh=[1,2,3])
y_pred = f.predict(X=X_pred)
print(y_pred)
```

**Reply by sakaiando** [reply]
This is very useful, thanks @ericjb!

**Reply by SultanOrazbayev** [reply]
Thanks, @ericjb, this is very useful! 

~~I still don't understand what exactly `DirectReductionForecaster` does that makes it behave differently even though `window_length` is set to zero.~~ (got explanation directly from Sakai)

**Reply by sakaiando** [reply]
@ericjb One more follow-up question if I may ask.

This leads to an error.
```
from sktime.forecasting.compose import YfromX
f = YfromX(LinearRegression(fit_intercept=False))
f.fit(y=y, X=X, fh=[1,2,3]).predict(X_pred)
```
But the following code that adds "X=" works. 
```
from sktime.forecasting.compose import YfromX
f = YfromX(LinearRegression(fit_intercept=False))
f.fit(y=y, X=X, fh=[1,2,3]).predict(X=X_pred)
```
Is this an expected behavior or a bug?

**Reply by SultanOrazbayev** [reply]
@sakaiando : not100% sure, but looks like the `.predict` expects arguments in order of `fh` first and `X` second, so unless X is specified explicitly, the code `.predict(X_pred)` attempts to pass X as `fh`.

**Reply by sakaiando** [reply]
Thanks @SultanOrazbayev, it would be great if YfromX documentation mentions it!
https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.forecasting.compose.YfromX.html
